### PR TITLE
fix(search): Fix search page to respond to searches made from sidebar

### LINKED ISSUE #6597: Search page doesn't respond when entering search terms from sidebar
## Expected Behavior
When you're on the search page, entering new terms from the sidebar search input box should update the search page

## Current Behavior

Updating search terms from sidebar doesn't update query param, search bar, or search results.

## Possible Solution
`SearchContext` and `SearchBar` components needs to respond to changes to initial search term prop

## Steps to Reproduce

1. Open Search plugin page
2. Enter some random search term on search bar
3. Enter a different search term from sidebar search input
4. Search page retains old search and doesn't update

## Context

If you're not on the search page and use the search on the sidebar then it correctly routes you to the search page and performs the requested search on the term. This is a obscure edge case where if you're already on the search page, the sidebar search doesn't work.

## Your Environment

- NodeJS Version (v12): 14
- Operating System and Version (e.g. Ubuntu 14.04): Ubuntu
- Browser Information: Chrome

### PR DESCRIPTION
**Pull request by kuangp**
Signed-off-by: Phil Kuang <pkuang@factset.com>

## Hey, I just made a Pull Request!

Closes #6597 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by timbonicus — APPROVED** [review]
Approving the changeset, will leave the code changes to @backstage/techdocs-core 🙂  If you're clever enough to spot a techdocs/search-only PR, you can prefix the changeset with [techdocs-](https://github.com/backstage/backstage/blob/master/.github/CODEOWNERS#L24) to avoid reviewers altogether.

**Review by iamEAP — COMMENTED** [review]
Thanks @kuangp!

It might be worth a second to step back and consider the intention behind some of these bits:

- The `<SearchContext />` is intended to be a situation-agnostic encapsulation for query state, performing searches, and then result state.
- The `<SearchBar />` is intended to be a situation-agnostic way to read/control the term on that `SearchContext`.
- The `<SearchPage />` is a specific situation in which the above is used where there happens to be an expectation that URL state and `SearchContext` state is synced in some way.
- The `<SidebarSearch />` is context-agnostic way to get to that search page.

My hope is that the fix for this can more or less be confined to the `<SearchPage />` component, since we want to keep the `<SearchBar />` and `<SearchContext />` very generic.

Perhaps what's needed instead is a slight revamp of the `UrlUpdater` in `<SearchPage />` that, rather than just quietly syncing search state updates one-way to the URL, actively manages search state based on _changes_ to the URL (though only in limited circumstances)

**Comment by kuangp** [comment]
@iamEAP Thanks for your insight! I took your suggestion and refactored `UrlUpdater` to more actively manage syncing search state with the URL. Let me know what you think!

**Comment by ottosichert** [comment]
Hey @kuangp, thanks for taking on this issue! I have checked out the latest version 1e1187362021ea342bb27a7c2b722e33a746725b and found following differences in search behaviour:

- By default, the `type` `software-catalog` was selected before, which seems to been gone with this change. Is this intended?
- Similarly, when typing any term in the side bar and confirming with `Enter`, the search page will have an empty search bar, both when being on the search page already or somewhere else

I'm on Firefox 90 and tried the above with a fresh `yarn install`, can you reproduce this behaviour?

**Comment by kuangp** [comment]
@ottosichert thanks for reporting these issues! I updated the change with a fix and confirmed it on Firefox.

File: plugins/search/src/components/SearchPage/SearchPage.test.tsx [hunk 1]
```diff
@@ -75,13 +84,11 @@ describe('SearchPage', () => {
...
 
-    // Then search context should be initialized with these values...
-    const calls = (SearchContextProvider as jest.Mock).mock.calls[0];
-    const actualInitialState = calls[0].initialState;
-    expect(actualInitialState.term).toEqual(expectedTerm);
-    expect(actualInitialState.types).toEqual(expectedTypes);
-    expect(actualInitialState.pageCursor).toEqual(expectedPageCursor);
-    expect(actualInitialState.filters).toStrictEqual(expectedFilters);
+    // Then search context should be set with these values...
+    expect(setTermMock.mock.calls[0][0]).toEqual(expectedTerm);
```
Comment (freben) [inline plugins/search/src/components/SearchPage/SearchPage.test.tsx:88]: sounds like these could be on the form `expect(mock).toHaveBeenCalledWith(value)` instead

**Review by freben — APPROVED** [review]
This is looking OK as far as I can tell! @backstage/techdocs-core Is this ready for merge?

**Comment by kuangp** [comment]
@iamEAP do you have any other comments on this change?

File: plugins/search/src/components/SearchBar/SearchBar.tsx [hunk 2]
```diff
@@ -31,6 +31,19 @@ export const SearchBar = ({ className, debounceTime = 0 }: Props) => {
...
+  // taking into account changes to valueRef triggering the hook
+  const valueRef = useRef(value);
+  useEffect(() => {
+    // Only respond to changes to term
+    if (valueRef.current === value && value !== term) {
+      setValue(term);
+    }
+    valueRef.current = value;
+  }, [term, value]);
+
```
Comment (camilaibs) [inline plugins/search/src/components/SearchBar/SearchBar.tsx:46]: We already have access to the previous value by passing a function to `setValue`, so testing locally it works without refs and the state will not be updated if it returns the previous value (else): 

```suggestion
  useEffect(() => {
    setValue(prevValue => prevValue !== term ? term : prevValue);
  }, [term]);
```

File: plugins/search/src/components/SearchPage/SearchPage.tsx [hunk 3]
```diff
@@ -14,54 +14,92 @@
...
+    term,
+    setTerm,
     types,
+    setTypes,
     pageCursor,
+    setPageCursor,
     filters,
-  };
+    setFilters,
+  ]);
```
Comment (camilaibs) [inline plugins/search/src/components/SearchPage/SearchPage.tsx:74]: We're not using term, types, pageCursor and filters inside useEffect, and this causes unnecessary calls

**Review by camilaibs — CHANGES_REQUESTED** [review]
Hi @kuangp, thanks so much for this fix, I just added 2 little comments.

File: plugins/search/src/components/SearchPage/SearchPage.tsx (hunk 3 above)
Comment (kuangp) [reply plugins/search/src/components/SearchPage/SearchPage.tsx:74]: thanks! forgot to remove these during my previous refactor

File: plugins/search/src/components/SearchBar/SearchBar.tsx (hunk 2 above)
Comment (kuangp) [reply plugins/search/src/components/SearchBar/SearchBar.tsx:46]: ah thanks for the tip! definitely looks a cleaner this way

**Comment by kuangp** [comment]
@camilaibs updated based off your suggestions, thanks!

**Review by camilaibs — APPROVED** [review]
Looks Good!

**Pull request merged**